### PR TITLE
Csweaver/cli

### PR DESCRIPTION
**Pull request by csweaver**
Stage 1 of CLI implemented.

Now you can run with:
cellxgene run [engine] [dataset directory]

### CODE REVIEW

File: server/app/app.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import os
+import argparse
```
Comment (ttung) [inline server/app/app.py]: keep alphabetized?

File: server/app/app.py [hunk 2]
```diff
@@ -53,5 +40,33 @@
 app.add_url_rule("/", endpoint="index")
 
 
+def run(args):
+    global data
```
Comment (ttung) [inline server/app/app.py]: generally not a fan of globals.

consider attaching the data object to the app object (possibly in a separate PR).

consider also instantiating the app object here (i.e., in `run()`), and converting any references to `server.app.app.app` to [flask.current_app](http://flask.pocoo.org/docs/1.0/api/#flask.current_app) in a separate PR.

File: server/app/app.py [hunk 3]
```diff
@@ -53,5 +40,33 @@
...
+    run_subparser.add_argument("--title", "-t", help="Title to display (default = data directory name")
+    run_subparser.add_argument("--port", help="Port to run server on.", type=int, default=5005)
+    run_subparser.add_argument("engine", help="The underlying structure of the data")
+    run_subparser.add_argument("data_directory", metavar="dir", help="Directory containing data and schema")
+    run_subparser.set_defaults(func=run)
+    args = parser.parse_args()
+
+    args.func(args)
+
+    # app.run(host="0.0.0.0", debug=True, port=5005)
```
Comment (ttung) [inline server/app/app.py]: remove.

File: server/app/app.py [hunk 4]
```diff
@@ -53,5 +40,33 @@
...
+        data = ScanpyEngine(args.data_directory, schema="data_schema.json")
+    app.run(host="0.0.0.0", debug=True, port=args.port)
+
+
 def main():
-    app.run(host="0.0.0.0", debug=True, port=5005)
+    parser = argparse.ArgumentParser(description="Cellxgene is a tool for exploring single cell expression.")
+    subparsers = parser.add_subparsers(dest="cellxgene_command")
+    run_subparser = subparsers.add_parser("run", help="run cellxgene")
+    run_subparser.add_argument("--title", "-t", help="Title to display (default = data directory name")
```
Comment (ttung) [inline server/app/app.py]: I had to dig into the code to figure out what this meant.  Consider clarifying.

File: server/app/app.py [hunk 5]
```diff
@@ -53,5 +40,33 @@
...
+    run_subparser.add_argument("--port", help="Port to run server on.", type=int, default=5005)
+    run_subparser.add_argument("engine", help="The underlying structure of the data")
```
Comment (ttung) [inline server/app/app.py]: Is this right?

File: server/app/app.py (hunk 1 above)
Comment (csweaver) [reply server/app/app.py]: sure, I'll fix

File: server/app/app.py (hunk 2 above)
Comment (csweaver) [reply server/app/app.py]: I didn't like it either, I prefer your idea. I wasn't quite sure what the best alternative to be and wanted to keep this PR relatively small. I like your idea, let's discuss later today.

File: server/app/app.py (hunk 3 above)
Comment (csweaver) [reply server/app/app.py]: ok

File: server/app/app.py (hunk 4 above)
Comment (csweaver) [reply server/app/app.py]: Which part? What the run command means?

File: server/app/app.py (hunk 5 above)
Comment (csweaver) [reply server/app/app.py]: Is which right?

File: server/app/app.py (hunk 4 above)
Comment (ttung) [reply server/app/app.py]: Where the title is shown.

Comment (csweaver) [reply server/app/app.py]: NP, I'll write some better help text

File: server/app/app.py (hunk 5 above)
Comment (ttung) [reply server/app/app.py]: I thought engine was EM2 vs etc.

Comment (csweaver) [reply server/app/app.py]: yeah, I can write a better help text for that. I'm not quite sure how to describe it best in the help text.

Comment (ttung) [reply server/app/app.py]: If you make it a [choice](https://docs.python.org/3/library/argparse.html#choices), it might be pretty self-describing.

Comment (csweaver) [reply server/app/app.py]: Shifted from `cellxgene run <engine>` to `cellxgene <engine>` I think that make this a little clearer.

**Comment by ttung** [comment]
This now has the changes from https://github.com/chanzuckerberg/cellxgene/pull/138.  i think you may need to rebase.

**Pull request merged**